http4k-server/websocket/src/main/kotlin/org/http4k/server/websocket/JavaWebSocket.kt

```diff
@@ -0,0 +1,135 @@
+import java.util.concurrent.TimeUnit
+
+class JavaWebSocket(
+    private val port: Int = 8000,
+    private val hostName: String = "0.0.0.0",
+    private val drafts: List<Draft>? = null,
+    override val stopMode: ServerConfig.StopMode = ServerConfig.StopMode.Immediate,
+    private val addShutdownHook: Boolean = true,
+    private val startupTimeout: Duration = Duration.ofSeconds(5),
+    private val configFn: WebSocketServer.() -> Unit = {
+        isReuseAddr = true // Set SO_REUSEADDR by default
+    },
+): PolyServerConfig {
+
+    override fun toServer(http: HttpHandler?, ws: WsHandler?, sse: SseHandler?): Http4kServer {
+        if (http != null) throw UnsupportedOperationException("JavaWebSocket does not support http")
+        if (ws == null) throw IllegalStateException("JavaWebSocket requires a WsHandler")
+        if (sse != null) throw UnsupportedOperationException("JavaWebSocket does not support sse")
+
+        val startLatch = CountDownLatch(1)
+        val address = InetSocketAddress(hostName, port)
+        val server = createServer(ws, address, drafts, startLatch::countDown).also(configFn)
+
+        val http4kServer = object: Http4kServer {
+            override fun port() = server.port
+            override fun start() = also {
+                server.start()
+                startLatch.await(startupTimeout.toMillis(), TimeUnit.MILLISECONDS)
+            }
+            override fun stop() = also {
+                when (stopMode) {
+                    ServerConfig.StopMode.Immediate -> server.stop()
+                    is ServerConfig.StopMode.Graceful -> server.stop(stopMode.timeout.toMillis().toInt())
+                }
+            }
+        }
+
+        if (addShutdownHook) {
+            Runtime.getRuntime().addShutdownHook(Thread(http4kServer::stop))
+        }
+
+        return http4kServer
+    }
+}
+
+private fun createServer(
+    wsHandler: WsHandler,
+    address: InetSocketAddress,
+    drafts: List<Draft>?,
+    onServerStart: () -> Unit
+) = object: WebSocketServer(address, drafts) {
+
+    override fun onOpen(conn: WebSocket, handshake: ClientHandshake) {
+        val headers = handshake.iterateHttpFields()
+            .asSequence()
+            .map { it to handshake.getFieldValue(it) }
+            .toList()
+
+        val upgradeRequest = Request(Method.GET, handshake.resourceDescriptor)
+            .headers(headers)
+            .let { if (handshake.content != null) it.body(MemoryBody(handshake.content)) else it }
+            .source(RequestSource(conn.remoteSocketAddress.hostString, conn.remoteSocketAddress.port))
+
+        val wsAdapter = object : PushPullAdaptingWebSocket(upgradeRequest) {
+            override fun send(message: WsMessage) {
+                when (message.body) {
+                    is StreamBody -> conn.send(message.body.payload)
+                    else -> conn.send(message.bodyString())  // furthering the generalization that a MemoryBody is ALWAYS to use text mode
```

Comment:
This has been an annoyance for me in the past.  I use a binary message format for some of my applications, and that means I must always use a `StreamBody`, or else it gets sent in text mode and corrupted.  Is this intentional?

Reply:
TBH, I don't recall. 😂 ! If it was me, it probably seemed like the right thing to do at the time.

Reply:
Well, I don't want to break the convention here.  But might be worth looking into later.